aten/src/ATen/native/MathBitsFallback.h

```diff
@@ -72,55 +134,61 @@ struct MathOpFallback {
       return;
     }
 
-    // Mutable inputs to be tracked separately
-    std::vector<Tensor> mutable_inputs;
-
+    // updates for non-mutable inputs
+    bool check_for_alias_with_mut_arg = !mutable_inputs.empty();
     for (const auto i : c10::irange(num_arguments)) {
       auto& ivalue = (*stack)[stack_start + i];
       if (!(ivalue.isTensor() || ivalue.isTensorList())) {
         continue;
       }
       const auto& argument = arguments[i];
-      bool mut_arg = false;
       if (argument.alias_info()) {
         // Was already tested by is_write loop above
         TORCH_INTERNAL_ASSERT_DEBUG_ONLY(argument.alias_info()->isWrite());
-        mut_arg = true;
+        continue;
       }
       if (ivalue.isTensor()) {
-        if (!is_bit_set(ivalue.toTensor())) {
-          continue;
-        }
-
         auto tensor = std::move(ivalue).toTensor();
         TORCH_CHECK_NOT_IMPLEMENTED(!tensor.is_meta(), op_name, " fallback does not support meta tensors.");
-        if (mut_arg) {
-          // TODO: This is a waste if the argument is write only
-          _set_bit(tensor, false);
-          math_op_(tensor);
-          mutable_inputs.emplace_back(tensor);
-        } else {
+        bool resolve_needed = true;
+        if (check_for_alias_with_mut_arg) {
+          for (const auto& mutable_input : mutable_inputs) {
+            // check if tensor shares memory with one of the mutable tensors
+            // with math bit set to True.
+            // This check is crucial since we in-place materialize the bit for the mutable tensor(s) which
+            // might be sharing memory with one of the non-mutable tensors that could be reading into the wrong
+            // values if their math bit is not set to True.
+            // no op if tensor and mutable arg share memory and have math bit set to True
+            if (tensor.is_alias_of(mutable_input)) {
```

Comment:
What is this function checking for? Full aliasing like they look at the exactly the same memory but not a subset?
Partial overlaps will just get ignored here and materialized out of place because resolve_needed remains "true"?

Reply:
synced offline: check for memory overlap using https://github.com/pytorch/pytorch/blob/839eaa2e91556ecd4532596b4fef18a1c3f6e1c1/aten/src/ATen/MemoryOverlap.cpp#L39-L72